### PR TITLE
feat(vscode): Added Dynamic C# File Generation and Unit Test SDK Support

### PR DESCRIPTION
**Pull request by samikay101**
## Requirement Checklist

* [x] The commit message follows our guidelines
* [x] Tests for the changes have been added (for bug fixes or features)

## Type of Change

* [ ] Bug fix
* [x] Feature
* [ ] Other

## Current Behavior

The current implementation does not support a floating version for the `Microsoft.Azure.Workflows.WebJobs.Extension` package in the project files. Users are required to specify exact versions, which limits flexibility when it comes to updates.

Additionally, placeholders like `<%= LogicAppName %>`, `<%= UnitTestName %>`, and `<%= WorkflowName %>` in the C# unit test templates were not dynamically replaced with actual values. This required manual changes to the generated files.

The project file also did not yet include the necessary references to the unit test SDK that is being developed for backend unit testing.

## New Behavior

- Implemented wildcard support for the `Microsoft.Azure.Workflows.WebJobs.Extension` package, allowing the use of floating versions like `17.10.*`. This ensures that the highest available version within the range is automatically resolved.

- Dynamically replaced placeholders like `<%= LogicAppName %>`, `<%= UnitTestName %>`, and `<%= WorkflowName %>` in the C# unit test files with actual values during generation. The `createCsFile` method handles this by reading the template, performing the replacement, and writing the final file.

- Updated the project file to include the necessary dependencies for the unit test SDK being developed on the backend for unit testing purposes. [Link to work item for unit test SDK support](https://msazure.visualstudio.com/One/_workitems/edit/29147348)

- Cleaned up the UX for the unit test generation in the following ways:
  1. Hidden the "Create unit test from scratch" button.
  2. Removed wiring for the unit test edit and management from the codeless approach to better align with the SDK-based approach.
  3. Ensured the extension no longer opens the designer to author unit tests, avoiding the previous codeless editing behavior.
  4. Cleaned up UX gestures for the codeless experience in VS Code, including unlinking test explorer initialization on extension startup and hiding unnecessary elements. [Link to work item for cleaned up UX](https://msazure.visualstudio.com/One/_workitems/edit/29299788)

## Impact of Change

* [ ] **This is a breaking change.**

This update should not introduce any breaking changes for downstream consumers. It only adds flexibility with floating version support, introduces dynamic template replacements in C# unit test generation, and improves the user interface.

### CODE REVIEW

File: apps/vs-code-designer/src/app/commands/workflows/unitTest/createUnitTest.ts [hunk 1]
```diff
@@ -154,7 +154,7 @@ async function generateCodefulUnitTest(
 
     ext.outputChannel.appendLog(localize('filesUnzipped', 'Files successfully unzipped.'));
     context.telemetry.properties.processStage = 'Files Unzipped';
```
Comment (ecfan) [inline apps/vs-code-designer/src/app/commands/workflows/unitTest/createUnitTest.ts:156]:
```suggestion
    context.telemetry.properties.processStage = 'Files unzipped';
```

File: apps/vs-code-designer/src/main.ts [hunk 2]
```diff
@@ -67,7 +66,8 @@ export async function activate(context: vscode.ExtensionContext) {
     promptParameterizeConnections(activateContext);
     verifyLocalConnectionKeys(activateContext);
     await startOnboarding(activateContext);
-    await prepareTestExplorer(context, activateContext);
+    // comment out test explorer and removed import for codeful approach
```
Comment (ecfan) [inline apps/vs-code-designer/src/main.ts:69]:
```suggestion
    // Comment out Test Explorer and removed import for codeful approach
```

File: apps/vs-code-designer/src/main.ts [hunk 3]
```diff
@@ -67,7 +66,8 @@ export async function activate(context: vscode.ExtensionContext) {
...
+    //await prepareTestExplorer(context, activateContext);
```
Comment (ecfan) [inline apps/vs-code-designer/src/main.ts:70]:
```suggestion
    // await prepareTestExplorer(context, activateContext);
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            // {
+            //     'statusCode': '200',
+            //     'body': {
+            //         <Your Test JSON Value>
+            //     }
+            // }");
+
+            // Sample mock data 2: Modify existing mocked data for <actionName> dynamically.
+            // var mockDataToModify = this.actionMocks["<actionName>"];
+            // mockDataToModify.Outputs["body"]["<property1Name>"] = "<Your Test String Value>";
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:115]:
```suggestion
            // mockDataToModify.Outputs["body"]["<property1Name>"] = "<your-test-string-value>";
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            // Sample mock data 1: Replace <Your Test JSON Value> with the appropriate mock data for the test.
+            // return JObject.Parse(@"
+            // {
+            //     'statusCode': '200',
+            //     'body': {
+            //         <Your Test JSON Value>
+            //     }
+            // }");
+
+            // Sample mock data 2: Modify existing mocked data for <actionName> dynamically.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:113]:
```suggestion
            // Sample mock data 2: Modify the existing mocked data dynamically for <actionName>.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        /// <param name="context">The test execution context that contains information about the current test run.</param>
+        public JToken Mock<actionName>OutputCallback(TestExecutionContext context)
+        {
+            // Sample mock data 1: Replace <Your Test JSON Value> with the appropriate mock data for the test.
+            // return JObject.Parse(@"
+            // {
+            //     'statusCode': '200',
+            //     'body': {
+            //         <Your Test JSON Value>
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:109]:
```suggestion
            //         <your-test-JSON value>
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        #region Mock generator helpers
+
+        /// <summary>
+        /// Callback method to dynamically generate mocked data for the action '<actionName>'.
+        /// This method can be modified to return different mock outputs based on the test scenario.
+        /// </summary>
+        /// <param name="context">The test execution context that contains information about the current test run.</param>
+        public JToken Mock<actionName>OutputCallback(TestExecutionContext context)
+        {
+            // Sample mock data 1: Replace <Your Test JSON Value> with the appropriate mock data for the test.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:104]:
```suggestion
            // Sample mock data 1: Replace <your-test-JSON-value> with the appropriate mock data for the test.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            // Verify that the workflow executed successfully and that the status is 'Succeeded'.
+            Assert.IsNotNull(testRun);
+            Assert.AreEqual(TestWorkflowStatus.Succeeded, testRun.Status);
+        }
+
+        #region Mock generator helpers
+
+        /// <summary>
+        /// Callback method to dynamically generate mocked data for the action '<actionName>'.
+        /// This method can be modified to return different mock outputs based on the test scenario.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:99]:
```suggestion
        /// You can modify this method to return different mock outputs based on the test scenario.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            // ASSERT
+            // Verify that the workflow executed successfully and that the status is 'Succeeded'.
+            Assert.IsNotNull(testRun);
+            Assert.AreEqual(TestWorkflowStatus.Succeeded, testRun.Status);
+        }
+
+        #region Mock generator helpers
+
+        /// <summary>
+        /// Callback method to dynamically generate mocked data for the action '<actionName>'.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:98]:
```suggestion
        /// The callback method to dynamically generate mocked data for the action named '<actionName>'.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+                onGetActionOutputsCallback: this.Mock<actionName>OutputCallback
+            );
+
+            // ACT 
+            // Create an instance of UnitTestExecutor and run the workflow with the mock data.
+            var executor = new UnitTestExecutor(this.workflowDefinitionPath, this.connectionsPath, this.parametersPath, this.localSettingsPath);
+            var testRun = await executor.RunWorkflowAsync(this.triggerMock, actionMocks).ConfigureAwait(false);
+
+            // ASSERT
+            // Verify that the workflow executed successfully and that the status is 'Succeeded'.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:90]:
```suggestion
            // Verify that the workflow executed successfully, and the status is 'Succeeded'.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+            // Load the mock data from the <%= UnitTestName %>-mock.json file.
+            var mockData = JObject.Parse(File.ReadAllText(@"<%= UnitTestName %>-mock.json"));
+            this.triggerMock = mockData["triggerMocks"].ToObject<Dictionary<string, TriggerMock>>().FirstOrDefault().Value;
+            this.actionMocks = mockData["actionMocks"].ToObject<Dictionary<string, ActionMock>>();
+        }
+
+        /// <summary>
+        /// Sample unit test for executing the <%= WorkflowName %> workflow.
+        /// This method demonstrates how to set up mocks, execute the workflow, and assert the outcome.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:70]:
```suggestion
        /// This method shows how to set up mocks, execute the workflow, and assert the outcome.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+            this.localSettingsPath = @"../<%= logicAppName %>/local.settings.json";
+
+            // Load the mock data from the <%= UnitTestName %>-mock.json file.
+            var mockData = JObject.Parse(File.ReadAllText(@"<%= UnitTestName %>-mock.json"));
+            this.triggerMock = mockData["triggerMocks"].ToObject<Dictionary<string, TriggerMock>>().FirstOrDefault().Value;
+            this.actionMocks = mockData["actionMocks"].ToObject<Dictionary<string, ActionMock>>();
+        }
+
+        /// <summary>
+        /// Sample unit test for executing the <%= WorkflowName %> workflow.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:69]:
```suggestion
        /// A sample unit test for executing the workflow named <%= WorkflowName %>.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        /// It also loads the mock data from the <%= UnitTestName %>-mock.json file.
+        /// </summary>
+        public <%= UnitTestName %>()
+        {
+            this.workflowDefinitionPath = @"../<%= logicAppName %>/<%= workflowName %>/workflow.json";
+            this.connectionsPath = @"../<%= logicAppName %>/connections.json";
+            this.parametersPath = @"../<%= logicAppName %>/parameters.json";
+            this.localSettingsPath = @"../<%= logicAppName %>/local.settings.json";
+
+            // Load the mock data from the <%= UnitTestName %>-mock.json file.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:62]:
```suggestion
            // Load the mock data from the file named <%= UnitTestName %>-mock.json.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+
+        /// <summary>
+        /// The mocked data for actions generated from a workflow run.
+        /// </summary>
+        public Dictionary<string, ActionMock> actionMocks;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="<%= UnitTestName %>" /> class.
+        /// This constructor loads the necessary paths for the workflow, connections, parameters, and local settings JSON files.
+        /// It also loads the mock data from the <%= UnitTestName %>-mock.json file.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:53]:
```suggestion
        /// The constructor also loads the mock data from the file named <%= UnitTestName %>-mock.json.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        public TriggerMock triggerMock;
+
+        /// <summary>
+        /// The mocked data for actions generated from a workflow run.
+        /// </summary>
+        public Dictionary<string, ActionMock> actionMocks;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="<%= UnitTestName %>" /> class.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:51]:
```suggestion
        /// Initializes a new instance of the class named <see cref="<%= UnitTestName %>" />.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        private string localSettingsPath;
+
+        /// <summary>
+        /// The mocked data for the trigger generated from a workflow run.
+        /// </summary>
+        public TriggerMock triggerMock;
+
+        /// <summary>
+        /// The mocked data for actions generated from a workflow run.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:46]: What does "generated" describe here, the mocked data or the actions?

```suggestion
        /// The mocked data for the actions generated from a workflow run.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        private string parametersPath;
+
+        /// <summary>
+        /// The relative path to local settings JSON file.
+        /// </summary>
+        private string localSettingsPath;
+
+        /// <summary>
+        /// The mocked data for the trigger generated from a workflow run.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:41]: What does "generated" describe here, the mocked data or the trigger?

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    using Microsoft.Azure.Workflows.UnitTesting.Definitions;
+    using Newtonsoft.Json.Linq;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Unit test class for testing the <%= WorkflowName %> workflow.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:13]:
```suggestion
    /// The unit test class for testing the workflow named <%= WorkflowName %>.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+    /// This class is auto-generated when the user selects to create a unit test from a workflow run in VS Code.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:14]: We never abbreviate "Visual Studio":

```suggestion
    /// This class is automatically generated when you choose to create a unit test from a workflow run in Visual Studio Code.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+    /// The user can edit this class to modify the test, mock data, and assertions as necessary.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:15]:
```suggestion
    /// You can edit this class to modify the test, mock data, and assertions as necessary.
```

**Review by ecfan — CHANGES_REQUESTED** [review]
@samikay101: Feedback noted, thanks!

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile (hunk 17 above)
Comment (samikay101) [reply apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:41]: The data refers to the output of the actions from the run, which is being mocked through the generation of the C# file.

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile (hunk 16 above)
Comment (samikay101) [reply apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:46]: The data refers to the output of the actions from the run, which is mocked in the mock.json file and referenced in the C# file.

**Comment by samikay101** [comment]
@ecfan I’ve implemented your suggestions and responded to your questions. Please let me know if you have any additional feedback or comments.

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile (hunk 17 above)
Comment (ecfan) [reply apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:41]: Thanks for the clarification, here's my suggestion:

```suggestion
        /// The mocked trigger outputs, which are generated from a workflow run.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+    /// <summary>
+    /// The unit test class for testing the workflow named <%= WorkflowName %>.
+    /// This class is automatically generated when you choose to create a unit test from a workflow run in Visual Studio Code.
+    /// You can edit this class to modify the test, mock data, and assertions as necessary.
+    /// </summary>
+    [TestClass]
+    public class <%= UnitTestName %>
+    {
+        /// <summary>
+        /// The relative path to workflow JSON file.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:21]:
```suggestion
        /// The relative path to the workflow JSON file.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        private string workflowDefinitionPath;
+
+        /// <summary>
+        /// The relative path to connections JSON file.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:26]:
```suggestion
        /// The relative path to the connections JSON file.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        private string connectionsPath;
+
+        /// <summary>
+        /// The relative path to parameters JSON file.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:31]:
```suggestion
        /// The relative path to the parameters JSON file.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        private string parametersPath;
+
+        /// <summary>
+        /// The relative path to local settings JSON file.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:36]:
```suggestion
        /// The relative path to the local settings JSON file.
```

File: apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+        /// </summary>
+        private string localSettingsPath;
+
+        /// <summary>
+        /// The mocked data for the trigger generated from a workflow run.
+        /// </summary>
+        public TriggerMock triggerMock;
+
+        /// <summary>
+        /// The mocked data for the actions generated from a workflow run.
```
Comment (ecfan) [inline apps/vs-code-designer/dist/assets/UnitTestTemplates/TestClassFile:46]:
```suggestion
        /// The mocked actions' outputs that are generated from a workflow run.
```

**Review by ecfan — APPROVED** [review]

**Review by ccastrotrejo — APPROVED** [review]

**Review by ecfan — APPROVED** [review]

**Pull request merged**